### DISCUSSION TITLE
Unconsumed column names when inserting into remote database

### DISCUSSION THREAD

**Discussion by LittleBunnyFooBar** [sqlalchemy/sqlalchemy#7456] [Usage Questions]
### Describe the bug

I am loading my database via reflection, searching for additional data using an API, and attempting to insert any results found into the same database. I can successfully update a record, but am receiving an "unconsumed column names" error when attempting to insert a new record.

### To Reproduce

```python
import sqlalchemy
import requests

from sshtunnel import SSHTunnelForwarder
from sqlalchemy import create_engine
from sqlalchemy import Table, MetaData
from rdkit import Chem

if __name__ == '__main__':
    with SSHTunnelForwarder(
            ('...', 22),
            ssh_username="uname",
            ssh_password="passwd",
            remote_bind_address=('localhost', 5432)) as server:

        server.start()
        print('Server connected via SSH')

        # connect to PostgreSQL
        local_port                  = str(server.local_bind_port)
        engine                      = create_engine('postgresql://uname:passwd@127.0.0.1:' + local_port + '/sandbox')
        connection                  = engine.connect()
        metadata                    = MetaData(engine)
        alias                       = Table('alias', metadata, autoload_with=engine)
        known_biology               = Table('known_biology', metadata, autload_with=engine)

        print('Database connection opened')
        query = sqlalchemy.text("""SELECT DISTINCT STUFF""")
        data = connection.execute(query)
        for row in data:
            se_id               = row['se_id']
            # Get MOL file from JChem
            structure_query = sqlalchemy.text("""
                SELECT cd_id, cd_structure, cd_smiles FROM jchem
                WHERE cd_id = :cd_id
            """)
            params      = {'cd_id' : row['se_cd_id']}
            structure   = connection.execute(structure_query, params).fetchone()
            if structure is not None:
                mol                 = bytes(structure["cd_structure"])
                mol_block           = Chem.MolFromMolBlock(mol)
                structure_string    = Chem.MolToInchiKey(mol_block)
                response            = requests.get('https://www.ebi.ac.uk/chembl/api/data/similarity/' + structure_string + '/100?format=json')
                # Skip if 404
                if response.status_code != 404:
                    response = response.json()
                    # Check if results are empty
                    if 'molecules' in response:
                        molecules       = response['molecules']
                        for molecule in molecules:
                            chembl_id    = molecule['molecule_chembl_id']
                            moa_response = requests.get('https://www.ebi.ac.uk/chembl/api/data/mechanism?molecule_chembl_id=' + chembl_id + '&format=json')
                            moa_response = moa_response.json()
                            for mechanism in moa_response['mechanisms']:
                                moa = mechanism['mechanism_of_action']
                                moa_insert =  known_biology.insert().values(sekb_se_id=se_id,sekb_skbt_id='1',sekb_description=moa,sekb_source='ChEMBL API - Mechanism of Action',sekb_ins_by='1')
                                connection.execute(moa_insert)
                else:
                    print("No ChEMBL CID: " + str(se_id))
            else:
                print("No MOL file found: " + str(se_id))
        # close the connection
        connection.close()
        print('Database connection closed')
```

### Error

Traceback (most recent call last):
  File "/mnt/c/Development/pycharm/chembl_api/main.py", line 65, in <module>
    connection.execute(moa_insert)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1289, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1473, in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 520, in _compile_w_cache
    compiled_sql = self._compiler(
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 555, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 776, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 451, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 486, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3795, in visit_insert
    crud_params = crud._get_crud_params(
  File "/home/jldavies/anaconda3/lib/python3.9/site-packages/sqlalchemy/sql/crud.py", line 168, in _get_crud_params
    raise exc.CompileError(
sqlalchemy.exc.CompileError: Unconsumed column names: sekb_se_id, sekb_skbt_id, sekb_description, sekb_ins_by, sekb_source

Process finished with exit code 1

### Versions

- OS: Windows 10 (WSL Debian)
- Python: 3.9
- SQLAlchemy: 1.4.28
- Database: postgresql 12.6
- DBAPI: psycopg2

**Comment by zzzeek** [comment]
hey there -

this error means that the keys you are passing to insert().values() don't match what the names of the columns are within the Table:

    moa_insert =  known_biology.insert().values(sekb_se_id=se_id,sekb_skbt_id='1',sekb_description=moa,sekb_source='ChEMBL API - Mechanism of Action',sekb_ins_by='1')

the error refers to all those names as "unconsumed" meaning when it renders an INSERT, it knows nothing about those names because they are not part of the .c. collection of the "known_biology" table.

Since you are using table reflection I can't see what the actual column names are, but you can what reflection is coming up with by printing out "known_biology.c.keys()".

**Reply by LittleBunnyFooBar** [reply]
Hey there

Unfortunately they are indeed a 1:1 match with my table column names. I wish that was the error. 

So this is interesting; when I do known_biology.c.keys(), the list is empty. So even though I can accurately query all of the other tables within my database, it cannot see this particular table.  Upon further inspection, I can only see 9 of my 50+ tables. 

![image](https://user-images.githubusercontent.com/3935241/145908883-8bc67406-c4f0-4fd0-8d2b-703bc285d84a.png)

**Comment by LittleBunnyFooBar** [comment]
Interesting! It looks like the following change fixes it:
```
metadata                    = MetaData()
metadata.reflect(bind=engine)
```

**Reply by zzzeek** [reply]
hard to tell what's happening, this line should be loading up all the columns:

    known_biology               = Table('known_biology', metadata, autload_with=engine)

**Reply by LittleBunnyFooBar** [reply]
@zzzeek yes, it seems bizarre that that wasn't loading up all of the tables. At least there's now a record of the workaround if it ever rears its head again.

**Comment by Romans-Sleepwalking** [comment]
```
moa_insert =  known_biology.insert().values(...)
```
You should insert the table instance as "insert" method parameter
```
moa_insert =  sqlalchemy.insert(known_biology).values(...)
```
